feat: add experimental http2 support

mafredri opened the pull request:
This change adds experimental support for WebSockets over HTTP/2 ([RFC 8441](https://www.rfc-editor.org/rfc/rfc8441.html)).

Implementation notes:

- To avoid breaking any flows, HTTP/2 functionality is currently opt-in on both server and client
- The HTTP/2 tests are currently kept in `internal/thirdparty/http2` (they require importing `golang.org/x/net/http2`, see https://github.com/golang/go/issues/67810)
- HTTP/2 extended CONNECT must currently be enabled via `GODEBUG`, see https://github.com/golang/go/issues/53208
- Explicit selection of HTTP/1 or HTTP/2 is required during `Dial` because we cannot test the underlying transport for support (for instance, `http.Transport` does not allow setting the HTTP/2 pseudo-header `:protocol`) 
- To `Dial` with HTTP/2 support, a `http2.Transport` must be provided via `DialOptions.Client` (since we do not import `golang.org/x/net/http2`)
- There are a few changes and cleanups unrelated to this feature, if needed they can be broken out into a separate branch

Closes #4

---

This change does not yet include benchmark tests for HTTP/2, those will be added at a later date as a follow-up.

mafredri committed:
feat: add experimental http2 support

Closes #4

zedkipp commented on .github/workflows/ci.yml:
```diff
@@ -17,7 +17,10 @@ jobs:
       - uses: actions/setup-go@v5
         with:
-          go-version-file: ./go.mod
+          # HACK(mafredri): The exampels and thirdparty library require Go 1.24
```
typo here and on line 33
```suggestion
          # HACK(mafredri): The examples and thirdparty library require Go 1.24
```

code-asher reviewed (approved):
Very nice 🤌  I messed around with the example as well and it all looks good.

mtojek reviewed (approved):
Amazing skills 👍 👍 

Have you tried to establish a connection using other tools? JS possibly?

mafredri commented on internal/examples/http2/README.md:
````diff
@@ -0,0 +1,62 @@
+
+```console
+# Server.
+$ cd examples/http2
````
Actually yeah, good catch, we'll need to update the other examples too but I can do that in a separate PR.

mafredri commented on protocol.go:
```diff
@@ -0,0 +1,55 @@
+// an http2.Transport (from golang.org/x/net/http2).
+//
+// Experimental: This type is experimental and may change in the future.
+type Protocol int
```
Honestly I was a bit on the fence about this myself, I almost did `HTTPProtocol` but it's a mouth/eyefull with the PP. 😅

On Dial we do have HTTPClient, HTTPHeader, so it's not too  much of a stretch to do HTTPProtocol.

mafredri commented on ci/test.sh:
```diff
@@ -1,36 +1,73 @@
+coverpkgsjoined=$(IFS=,; echo "${coverpkgs[*]}")
+
+echo "++++ Running main tests"
+go test --bench=. --timeout=1h -cover -covermode=atomic -coverpkg="$coverpkgsjoined" -test.gocoverdir="$coverbase/race=0" "$@" ./...
```
It's a typical timeout in this project all around. I'll keep it like that but we can follow-up with a cleanup regarding that.

mafredri commented on internal/examples/http2/main.go:
```diff
@@ -0,0 +1,67 @@
+	switch cmd := fs.Arg(0); cmd {
+	case "server":
+		if err := serverMain(prog, fs.Args()[1:]); err != nil {
+			fmt.Fprintf(os.Stderr, "server: error: %v\n", err)
+			os.Exit(1)
```
Since `log/slog` is the new kid on the block and doesn't have a fatal (IIRC), I think the explicit exit is OK. Could change the `fmt` to slog though.

mafredri commented on internal/examples/http2/server.go:
```diff
@@ -0,0 +1,172 @@
+				_ = ww.Close()
+				return
+			}
+			if err = ww.Close(); err != nil {
```
If we do we'll need to wrap this in a function (for loop and defer doesn't play nice).

mafredri commented on internal/examples/http2/server.go:
```diff
@@ -0,0 +1,172 @@
+		}
+
+		selfSigned := ""
+		if *certFile == "" && *keyFile == "" {
```
I added the auto-generated self-signed cert to make testing the example easier, but we can enforce cert/key, I was on the fence about adding cert generation anyway 👍🏻

mafredri commented on accept.go:
```diff
@@ -125,104 +147,206 @@ func accept(w http.ResponseWriter, r *http.Request, opts *AcceptOptions) (_ *Con
-	}); ok {
-		ginWriter.WriteHeaderNow()
+	switch opts.Protocol {
+	case ProtocolHTTP2:
```
I did not know about this, I'll need to investigate. Thanks for the heads up 👍🏻

mafredri replied:
I looked into this. Server-side already works with native h2c since the code checks `r.ProtoMajor == 2`, which handles both approaches the same way. The `GODEBUG=http2xconnect=1` env is still needed though.

Client-side is the main issue. I tested Go 1.25rc1, and the native `http.Transport` still rejects the `:protocol` pseudo-header, so the `http2.Transport` from x/net is needed.

Looks like full stdlib support is targeted for Go 1.26 (https://github.com/golang/go/issues/67810), we can revisit then.

mafredri committed:
fix: address review feedback

- Fix 'exampels' typo in CI workflow comments
- Fix README paths to use 'internal/examples/http2'
- Change visibleAddr to display 0.0.0.0 for all-interface bindings

mafredri commented on dial.go:
```diff
@@ -240,6 +317,17 @@ func secWebSocketKey(rr io.Reader) (string, error) {
+}
+
+func verifyServerResponseH1(opts *DialOptions, copts *compressionOptions, secWebSocketKey string, resp *http.Response) (*compressionOptions, error) {
 	if resp.StatusCode != http.StatusSwitchingProtocols {
```
https://github.com/coder/websocket/pull/550

mafredri commented on ci/static.sh:
```diff
@@ -0,0 +1,20 @@
+mkdir -p ./ci/out/static
+cp ./ci/out/coverage.html ./ci/out/static/coverage.html
+percent=$(go tool cover -func ./ci/out/profile.txt | tail -n1 | awk '{print $3}' | tr -d '%')
+wget -O ./ci/out/static/coverage.svg "https://img.shields.io/badge/coverage-${percent}%25-success"
```
https://github.com/coder/websocket/pull/549

mafredri committed:
refactor: rename Protocol to HTTPProtocol

Rename the Protocol type to HTTPProtocol for consistency with existing
HTTPClient and HTTPHeader field naming in DialOptions.

Constants are also renamed:
- ProtocolHTTP1 → HTTPProtocol1
- ProtocolHTTP2 → HTTPProtocol2
- ProtocolAcceptAny → HTTPProtocolAny

mafredri committed:
refactor(internal/examples/http2): remove self-signed cert generation

Simplify the TLS example by requiring -cert and -key flags instead of
auto-generating a self-signed certificate. This reduces complexity and
keeps the example focused on HTTP/2 WebSocket usage.

mafredri commented on accept.go:
```diff
@@ -125,104 +147,206 @@ func accept(w http.ResponseWriter, r *http.Request, opts *AcceptOptions) (_ *Con
+	switch proto {
+	case HTTPProtocol2:
+		// Prepare response headers for H2 (no Connection/Upgrade).
+		w.Header().Set("Sec-WebSocket-Accept", secWebSocketAccept(key))
```
Thanks for catching this @ambyjkl!

Nek-12 commented:
Are there plans to ship this?
